### ISSUE TITLE
PHP 8.3 warnings due to get_class deprecation

### ISSUE THREAD

**Issue by sammarshallou** [catalyst/moodle-auth_saml2#849]
In PHP 8.3, this can show warnings due to get_class deprecation - you can no longer call get_class with no params, need to use self::class or static::class or get_class($this). 

The warnings are in:

SimpleSAML/Auth/Source.php

Lines 179/180
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/.extlib/simplesamlphp/src/SimpleSAML/Auth/Source.php#L179

There is another instance of the same problem in Simple.php, I don't know if this one ever actually shows up in our configuration, but it seems like it should probably be fixed too.

These problems are all fixed in current version SimpleSAML library so upgrading that would fix it. 

Unfortunately updating the library is a bit of a pain and I don't have time to do it now :(  I have done a local workaround by changing the lines to the versions in the latest SimpleSAML ([php83.patch](https://github.com/user-attachments/files/18006399/php83.patch) attached, uses OU syntax for marking custom changes) but probably the right solution is to upgrade the library to latest.

**Comment by danmarsden** [comment]
@sammarshallou any chance you could do some further testing on the new branch?  https://github.com/catalyst/moodle-auth_saml2/tree/MOODLE_404_STABLE

**Comment by sammarshallou** [comment]
Oh for... I'm very sorry I had completely missed that there is a 404 branch!! I thought I had looked at that dropdown earlier, but obviously I looked at in the kind of way that doesn't involve actually seeing anything...

I have checked the source and it includes the fixes to these lines so I will cose this as already fixed. Also will update our local version...

**Comment by danmarsden** [comment]
It only landed here yesterday and hasn't been tested well yet. There's also a slmplesaml update due next week that we want to land before updating readme and the default branch. But any testing feedback would be great

**Comment by danmarsden** [comment]
Getting my weeks confused.simplesqmlpho 2.3.5 landed earlier this week and we need to get that in here pretty soon I think.

**Comment by sammarshallou** [comment]
I updated to the current MOODLE_404 version of auth_saml2 on our test server and er... I can still log in so I guess it's working... :) Beyond that top quality testing, we do actually have quite a lot of Behat tests related to the way our login is setup so I will see if any of those fail overnight, or if anyone else complains that the test server is broken.

**Comment by sammarshallou** [comment]
Automated testing results look OK, but I had to do a workaround because the 'twig' library is included in SimpleSAMLphp within auth_saml2, and it is also within qtype_coderunner which is another 3rd party plugin we use. The version in auth_saml2 is newer. 

This was causing the PHPunit run to crash a fatal error.

I don't know why this is breaking now but didn't before - possibilities include that more unit tests were added so that they actually now exercise something that uses the twig library and didn't before, or else that something in simplesamlphp changed so that it uses twig now for something it didn't before, I guess.

This is the workaround but I haven't checked if coderunner is actually broken as a result. Can't think of a better way to solve it so far... (Also not sure it causes any practical problem on normal requests other than whole unit test run)

https://gist.github.com/sammarshallou/1c3199a2b0d80d2024585b939d4615b9
